department-of-veterans-affairs/caseflow
# Jruuuu/APPEALS-21357

Resolves [APPEALS-21357](https://jira.devops.va.gov/browse/APPEALS-21357)

# Description
This story creates the seed data for vbms_ext_claims with and without associations to High Level Review End Product Establishments or Supplemental End Product Establishments

## Acceptance Criteria
- [x] create vbms_ext_claim factory.
- [x] create VbmsExtClaim Class that part of the Seeds module
- [x] create 3 different cases of VbmsExtClaims. None Associated to EPE HLR or SLC, Associated to EPE HLR or SLC & in_sync, and Associated to EPE HLR or SLC & out_of_sync
- [x] add VbmsExtClaim class to Seeds Module
- [x] create make target that seeds vbms_ext_claims, high level reviews, supplemental claims, and End Product Establishments
- [x] create unit and feature Rspec test for vbms_ext_claims_spec.rb for the Seeds Module  

## Testing Plan
1. Go to [APPEALS-24065](https://jira.devops.va.gov/browse/APPEALS-24065)

### Steps
1.  run `make rest`
	1. Expected Result: Runs without Error
2. run `make seed-dbs`
	1. Expected Result:
		1. Runs without Error
3. run `make external-db-remove`
	1. Expected Result:
		1. runs without error
		2. runs command: 
			1. `bundle exec rails r db/scripts/external/remove_vbms_ext_claim_table.rb`
		3. removes `vbms_ext_claim_table` from database caseflow_certification_development
4.  run `make external-db-create`
	1. Expected Result:
		1. runs without error
		2. runs both commands: 
			1. `bundle exec rails r db/scripts/external/create_vbms_ext_claim_table.rb`
		3. creates `vbms_ext_claim_table` in database caseflow_certification_development
5. run `make external-db-create-test`
	1. Expected Result:
		1. runs without error
		2. runs command
			1. `bundle exec rails r -e test db/scripts/external/create_vbms_ext_claim_table.rb`
6. run `make one-test spec/seeds/vbms_ext_claim_spec.rb`
	1. Expected Result:
		1. Test Pass without failures
7.  run `make seed-vbms-ext-claim`
	1. Expected Result:
		1. runs without error
		2. Seeds New 325 vbms_ext_claims
			1. Confirm in rails console that VbmsExtClaim.count  = 325
				1. run `make c`
				2. run `VbmsExtClaim.count`
				3. 325
			2. Confirm in rails console that VbmsExtClaim.CLAIMANT_PERSON_ID  matches End_Product_Establishment.claimant_participant_id
			3.  run `VbmsExtClaim.last`
			4.  run `EndProductEstablishment.last`
			5. verify that CLAIMANT_PERSON_ID matches claimant_participant_id
			6. verify that EP_CODE matches modifier
			7. verify that TYPE_CODE matches code
		3. Verify that 3 Seeding Methods Ran correctly 
			1. Expected Result:
				1. in DBeaver Confirm Data is correct in caseflow_certification_development database for the 3 cases
					1. Open DBeaver > Connect to Postgres > Databases > caseflow_certification_development > Schemas > public > and open tables vbms_ext_claims and end_product_establishments
					2. CASE: create_vbms_ext_claims_with_no_end_product_establishment  
						1. Verify that VbmsExtClaim.CLAIM_ID with EP_CODE: NULL does not exist in EndProductEstablishment.reference_id
							1. for example: 
								1. if VbmsExtClaim.CLAIM_ID = 100001 then EndProductEstablishment.reference_id should not have  100001
					3. CASE: create_in_sync_epes_and_vbms_ext_claims 
						1. Find an EPE.sync_status of "CAN"
						2. Verify that the following are matching
							1. VEC."CLAIM_ID" == EPE.reference_id
								VEC."CLAIM_DATE" == EPE.claim_date
								VEC."EP_CODE" == EPE.modifier
								VEC."CLAIMANT_PERSON_ID" == EPE.claimant_participant_id
								VEC."TYPE_CODE" == EPE.code
								VEC."PAYEE_CODE" == EPE.payee_code
								VEC."ESTABLISHMENT_DATE" == EPE.established_at
							2. Verify that HLR or SLC EPE has the correct data associated with the VEC
								1. Find a VEC with an EP_CODE
								2. Find the matching VEC."CLAIMANT_PERSON_ID" == EPE.claimant_participant_id
								3. verify EPE.reference_id == VEC.claim_id
								4. confirm EPE.source_type is  HLR or SLC
								5. confirm veteran_file_number match EPE.veteran_file_number == HLR.veteran_file_number  or EPE.veteran_file_number == SLC.veteran_file_number 
						3. Verify that they are in_sync
							1. VEC.CLAIM_ID == EPE.reference_id
							2. VEC.LEVEL_STATUS_CODE == EPE.sync_status
								1. for example:
									1.  VEC.LEVEL_STATUS_CODE = "CAN" 
									2.  EPE.sync_status = "CAN"
					1. CASE: create_out_of_sync_epes_and_vbms_ext_claims
						1. Find an EPE.sync_status of "PEND"
						2. Verify that the following are matching
							1. VEC."CLAIM_ID" == EPE.reference_id
								VEC."CLAIM_DATE" == EPE.claim_date
								VEC."EP_CODE" == EPE.modifier
								VEC."CLAIMANT_PERSON_ID" == EPE.claimant_participant_id
								VEC."TYPE_CODE" == EPE.code
								VEC."PAYEE_CODE" == EPE.payee_code
								VEC."ESTABLISHMENT_DATE" == EPE.established_at
						3. Verify that HLR or SLC EPE has the correct data associated with the VEC
							1. Find a VEC with an EP_CODE
							2. Find the matching VEC."CLAIMANT_PERSON_ID" == EPE.claimant_participant_id
							3. verify EPE.reference_id == VEC.claim_id
							4. confirm EPE.source_type is  HLR or SLC
							5. confirm veteran_file_number match 
							EPE.veteran_file_number == HLR.veteran_file_number  or EPE.veteran_file_number == SLC.veteran_file_number 
						4.  Verify that they are out_of_sync
							1. Find an EPE.sync_status of "PEND"
							2. Find the EPE.reference_id
							3. Find the Matching VEC.CLAIM_ID
							4. Verify that the VEC.LEVEL_STATUS_CODE do not match EPE.sync_status
								1. for example:
									1. EPE.sync_status = "PENDING"
									2. VEC.LEVEL_STATUS_CODE = "CAN"




## Tests
### Test Coverage
Did you include any test coverage for your code? Check below:

### Code Climate
Your code does not add any new code climate offenses? If so why?
